File: CssMinifier.java
// CssMinifier.java
import java.io.*;
import java.util.regex.*;

public class CssMinifier {
    private boolean removeComments;
    private boolean preserveSpaces;
    private boolean removeSemicolon;
    private boolean shrinkColors;

    public CssMinifier(boolean removeComments, boolean preserveSpaces,
                       boolean removeSemicolon, boolean shrinkColors) {
        this.removeComments = removeComments;
        this.preserveSpaces = preserveSpaces;
        this.removeSemicolon = removeSemicolon;
        this.shrinkColors = shrinkColors;
    }

    public String minify(String css) {
        if (removeComments) {
            css = css.replaceAll("/\\*.*?\\*/", "");
        }

        String[] lines = css.split("\n");
        StringBuilder sb = new StringBuilder();
        for (String line : lines) {
            String trimmed = line.trim();
            if (!trimmed.isEmpty()) {
                sb.append(trimmed);
            }
        }
        css = sb.toString();

        if (!preserveSpaces) {
            css = css.replaceAll("\\s*([{}:;,>+~()])\\s*", "$1");
            css = css.replaceAll("\\s*,\\s*", ",");
        }

        if (removeSemicolon) {
            css = css.replaceAll(";}", "}");
        }

        if (shrinkColors) {
            css = css.replaceAll("#([0-9a-fA-F])\\1([0-9a-fA-F])\\2([0-9a-fA-F])\\3", "#$1$2$3");
            css = css.replaceAll("\\b0(px|em|rem|ex|ch|vw|vh|vmin|vmax|%)", "0");
            css = css.replaceAll("\\b0+\\.(\\d+)", ".$1");
        }

        css = css.replaceAll("\n\\s*\n", "\n");
        return css.trim();
    }

    public static void main(String[] args) throws IOException {
        String inputFile = null;
        String outputFile = null;
        boolean noComments = false;
        boolean preserveSpaces = false;
        boolean noSemicolon = false;
        boolean shrinkColors = false;

        for (int i = 0; i < args.length; i++) {
            if ("--no-comments".equals(args[i])) noComments = true;
            else if ("--preserve-spaces".equals(args[i])) preserveSpaces = true;
            else if ("--no-semicolon".equals(args[i])) noSemicolon = true;
            else if ("--shrink-colors".equals(args[i])) shrinkColors = true;
            else if ("-o".equals(args[i]) && i+1 < args.length) outputFile = args[++i];
            else if (inputFile == null) inputFile = args[i];
        }

        String content;
        if (inputFile != null) {
            try (BufferedReader reader = new BufferedReader(new FileReader(inputFile))) {
                StringBuilder sb = new StringBuilder();
                String line;
                while ((line = reader.readLine()) != null) {
                    sb.append(line).append("\n");
                }
                content = sb.toString();
            }
        } else {
            try (BufferedReader reader = new BufferedReader(new InputStreamReader(System.in))) {
                StringBuilder sb = new StringBuilder();
                String line;
                while ((line = reader.readLine()) != null) {
                    sb.append(line).append("\n");
                }
                content = sb.toString();
            }
        }

        CssMinifier minifier = new CssMinifier(!noComments, preserveSpaces, !noSemicolon, shrinkColors);
        String minified = minifier.minify(content);

        if (outputFile != null) {
            try (FileWriter fw = new FileWriter(outputFile)) {
                fw.write(minified);
            }
        } else {
            System.out.println(minified);
        }
    }
}

File: css_minifier.swift
// css_minifier.swift
import Foundation

class CSSMinifier {
    var removeComments: Bool
    var preserveSpaces: Bool
    var removeSemicolon: Bool
    var shrinkColors: Bool

    init(removeComments: Bool = true, preserveSpaces: Bool = false,
         removeSemicolon: Bool = true, shrinkColors: Bool = false) {
        self.removeComments = removeComments
        self.preserveSpaces = preserveSpaces
        self.removeSemicolon = removeSemicolon
        self.shrinkColors = shrinkColors
    }

    func minify(_ css: String) -> String {
        var result = css

        if removeComments {
            let regex = try! NSRegularExpression(pattern: "/\\*.*?\\*/", options: .dotMatchesLineSeparators)
            result = regex.stringByReplacingMatches(in: result, range: NSRange(result.startIndex..., in: result), withTemplate: "")
        }

        let lines = result.components(separatedBy: .newlines)
        let trimmedLines = lines.map { $0.trimmingCharacters(in: .whitespaces) }.filter { !$0.isEmpty }
        result = trimmedLines.joined()

        if !preserveSpaces {
            let regex = try! NSRegularExpression(pattern: "\\s*([{}:;,>+~()])\\s*")
            result = regex.stringByReplacingMatches(in: result, range: NSRange(result.startIndex..., in: result), withTemplate: "$1")
            let regex2 = try! NSRegularExpression(pattern: "\\s*,\\s*")
            result = regex2.stringByReplacingMatches(in: result, range: NSRange(result.startIndex..., in: result), withTemplate: ",")
        }

        if removeSemicolon {
            let regex = try! NSRegularExpression(pattern: ";}")
            result = regex.stringByReplacingMatches(in: result, range: NSRange(result.startIndex..., in: result), withTemplate: "}")
        }

        if shrinkColors {
            let regex = try! NSRegularExpression(pattern: "#([0-9a-fA-F])\\1([0-9a-fA-F])\\2([0-9a-fA-F])\\3")
            result = regex.stringByReplacingMatches(in: result, range: NSRange(result.startIndex..., in: result), withTemplate: "#$1$2$3")
            let regex2 = try! NSRegularExpression(pattern: "\\b0(px|em|rem|ex|ch|vw|vh|vmin|vmax|%)")
            result = regex2.stringByReplacingMatches(in: result, range: NSRange(result.startIndex..., in: result), withTemplate: "0")
            let regex3 = try! NSRegularExpression(pattern: "\\b0+\\.(\\d+)")
            result = regex3.stringByReplacingMatches(in: result, range: NSRange(result.startIndex..., in: result), withTemplate: ".$1")
        }

        let regex = try! NSRegularExpression(pattern: "\n\\s*\n")
        result = regex.stringByReplacingMatches(in: result, range: NSRange(result.startIndex..., in: result), withTemplate: "\n")
        return result.trimmingCharacters(in: .whitespacesAndNewlines)
    }
}

func main() {
    let args = CommandLine.arguments.dropFirst()
    var inputFile: String? = nil
    var outputFile: String? = nil
    var noComments = false
    var preserveSpaces = false
    var noSemicolon = false
    var shrinkColors = false

    var i = 0
    let argsArray = Array(args)
    while i < argsArray.count {
        switch argsArray[i] {
        case "--no-comments": noComments = true
        case "--preserve-spaces": preserveSpaces = true
        case "--no-semicolon": noSemicolon = true
        case "--shrink-colors": shrinkColors = true
        case "-o":
            if i+1 < argsArray.count { outputFile = argsArray[i+1]; i+=1 }
        default:
            if inputFile == nil { inputFile = argsArray[i] }
        }
        i+=1
    }

    let content: String
    if let input = inputFile {
        guard let data = try? String(contentsOfFile: input, encoding: .utf8) else {
            print("Error reading file", to: &stderr)
            exit(1)
        }
        content = data
    } else {
        var data = ""
        while let line = readLine() {
            data += line + "\n"
        }
        content = data
    }

    let minifier = CSSMinifier(
        removeComments: !noComments,
        preserveSpaces: preserveSpaces,
        removeSemicolon: !noSemicolon,
        shrinkColors: shrinkColors
    )
    let minified = minifier.minify(content)

    if let output = outputFile {
        try? minified.write(toFile: output, atomically: true, encoding: .utf8)
    } else {
        print(minified)
    }
}

main()
